sparse/dok.py

```diff
@@ -0,0 +1,147 @@
+import six
+
+import numpy as np
+
+# Zip with Python 2/3 compat
+# Consumes less memory than Py2 zip
+from six.moves import zip, range
+
+from numbers import Integral
+from collections import Iterable
+
+from .slicing import normalize_index
+from .utils import _zero_of_dtype
+
+try:  # Windows compatibility
+    int = long
+except NameError:
+    pass
+
+
+class DOK:
+    def __init__(self, shape, values=None, dtype=None):
+        from .coo import COO
+        self.dict = {}
+
+        if isinstance(shape, COO):
+            self.dtype = shape.dtype
+            self.shape = shape.shape
+
+            for c, d in zip(shape.coords.T, shape.data):
+                self.dict[tuple(c)] = d
+
+            return
+
+        if isinstance(shape, np.ndarray):
+            coords = np.nonzero(shape)
+            data = shape[coords]
+            self.shape = shape.shape
+
+            for c in zip(data, *coords):
+                c, d = c[0], c[1:]
+                self.dict[c] = d
+
+            return
+
+        self.dtype = dtype
+
+        if isinstance(shape, Integral):
+            shape = (int(shape),)
+        elif isinstance(shape, Iterable):
+            for l in shape:
+                if not isinstance(l, Integral) or int(l) < 0:
+                    raise ValueError('shape must be an iterable of non-negative integers.')
+
+            shape = self.shape = tuple(shape)
+
+        if not values:
+            values = {}
+
+        if isinstance(values, dict):
+            for c, d in six.iteritems(values):
+                self[c] = d
+        else:
+            raise ValueError('values must be a dict.')
+
+        self.shape = shape
+        if dtype:
+            self.dtype = dtype
+
+    @property
+    def ndim(self):
+        return len(self.shape)
+
+    @property
+    def nnz(self):
+        return len(self.dict)
+
+    def __getitem__(self, item):
+        item = normalize_index(item, self.shape)
+
+        for i in item:
+            if not isinstance(i, Integral):
+                raise IndexError('All indices must be integers'
+                                 ' when getting an item.')
+
+        if not len(item) == self.ndim:
+            raise IndexError('Can only get single elements.')
```

Comment:
Might want to include the item in the error message with `"expected key of length %d, got %s" % (self.ndim, str(item))`